Add layout styles from Post Content block to post editor

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->


## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Fixes #42911, #44110, #44208.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Gets layout classnames and styles from Post Content block so post editor layout can match the front end in all respects.

WIP

Todo:

* remove hardcoded layout styles from post editor (custom content widths aren't working yet because they're being overridden by those)
* make sure child block alignments are correct per Post Content layout type
* `useLayoutClasses`, `useLayoutStyles` should probably be experimental or unstable?
* add tests

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. In site editor, change layout settings of Post Content block
2. Check that new layout is reflected in the front end

## Screenshots or screencast <!-- if applicable -->

tellthemachines committed:
Add layout styles from Post Content block to post editor

andrewserong commented on packages/block-editor/README.md:
```diff
@@ -766,6 +766,28 @@ _Parameters_
 
+### useLayoutClasses
+
+Generates the utility classnames for the given blocks layout attributes.
+This method was primarily added to reintroduce classnames that were removed
+in the 5.9 release (<https://github.com/WordPress/gutenberg/issues/38719>), rather
+than providing an extensive list of all possible layout classes. The plan is to
+have the style engine generate a more extensive list of utility classnames which
+will then replace this method.
```
This feedback is probably low priority, but just jotting down some thoughts:

When this function was originally written, I believe we were thinking that the style engine might be responsible for more of the layout support that we do now (I think this function was written prior to the layout refactor landing). It might be worth us updating the description here, as I believe it probably makes sense for `useLayoutClasses` to be the entry point for getting layout classnames — if in some point in the future we do use the style engine for constructing the classnames, we might wind up doing it as an internal implementation detail, so we'd still probably want this as the main function for getting those classnames?

In terms of stability / API signature, the function currently accepts the layout object, and a layoutDefinitions object. However, in `useLayoutStyles` the signature is a bit different in that it accepts a block object and a selector. Would it be worth coming up with a signature for both of these hooks that is pretty much the same, to create some consistency there? I like the use of the block object in the latter — we might wind up needing to access other aspects of the block object in `useLayoutClasses` in the future, too, so it could be worth making some changes there?

tellthemachines replied:
Thanks for thinking this through! I'll update the description to remove references to hypothetical future style engine work 😅 

> Would it be worth coming up with a signature for both of these hooks that is pretty much the same, to create some consistency there?

We wouldn't need the selector for `useLayoutClasses` though, would we? Other than that I'm happy to change it to take the block object.

andrewserong replied:
> We wouldn't need the selector for useLayoutClasses though, would we?

Ah, good point. Yeah, I can't think of a circumstance where we'd need the selector for getting the classnames. We might have use for it eventually for where we _output_ the classnames, but not for retrieving them in the first place?

tellthemachines replied:
Yeah I don't think we'd ever need the selector in `useLayoutClasses`.

I'm also struggling to think of a use for the whole block in this function; we only really need the layout object. We could potentially change it to get the layout definitions from `useSetting( 'layout' )` instead of passing them as a parameter. Would that work everywhere?

In `useLayoutStyles` we need both `layout` and `styles` from attributes, as well as the block name to pass to the layout type's `useLayoutStyles` function, so it can work out whether the block skips serialisation or not.

It's not looking like we can make the signatures very consistent here 😅 

Regarding the doc comment for `useLayoutClasses`, could we simply shorten it to "Generates the utility classnames for the given blocks layout attributes.", or is there any additional info that might be useful here?

Also, I'm thinking we should make these experimental given Layout itself has that status.

andrewserong replied:
> we only really need the layout object

Right now that's definitely the case. My only thought was if we wind up eventually needing to use parts of the `spacing` object for potential future classnames (e.g. if we needed to add a `has-gap` classname further down the track, which I think was discussed a little in the spacing presets discussions), then it'd be easier if we already had the object there.

> We could potentially change it to get the layout definitions from useSetting( 'layout' ) instead of passing them as a parameter. Would that work everywhere?

That very well could. I think passing it around is probably a result of some of these functions originally being pure functions instead of React hooks? Now that they're hooks, it's probably better to keep the API signature simpler instead of passing stuff around 🤔

> "Generates the utility classnames for the given blocks layout attributes."

Sounds good to me!

> Also, I'm thinking we should make these experimental given Layout itself has that status.

Good idea, there's still plenty that we're moving around, so would be good to maintain that flexibility 👍

andrewserong commented on packages/block-editor/src/hooks/layout.js:
```diff
@@ -89,6 +89,24 @@ function useLayoutClasses( layout, layoutDefinitions ) {
 	return layoutClassnames;
 }
 
+export function useLayoutStyles( block = {}, selector ) {
```
Nice abstraction here!

andrewserong commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -195,11 +206,25 @@ export default function VisualEditor( { styles } ) {
-		'is-layout-flow': ! themeSupportsLayout,
-		'has-global-padding': useRootPaddingAwareAlignments,
-	} );
+	const templateBlocks = parse( templateContent );
+	const postContentBlock = findPostContent( templateBlocks );
+	const postContentLayoutClasses = useLayoutClasses(
+		postContentBlock?.attributes?.layout,
+		defaultLayout?.definitions
+	);
+
+	const blockListLayoutClass = classnames(
+		{
+			'is-layout-flow': ! themeSupportsLayout,
+			'wp-container-visual-editor': themeSupportsLayout,
+		},
+		postContentLayoutClasses
+	);
+
+	const postContentLayoutStyles = useLayoutStyles(
+		postContentBlock,
+		'.wp-container-visual-editor'
+	);
```
Will we ultimately want to wrap this in a `useMemo` so that we only run it when `templateContent` changes?

tellthemachines replied:
Good question! I never know when it's best to `useMemo` or not 😅  Another option would be to return `null` from useLayoutStyles if there's no block.

tellthemachines replied:
> Another option would be to return null from useLayoutStyles if there's no block.

Actually I don't think we can do that because it's using hooks underneath 😕

andrewserong replied:
> I never know when it's best to useMemo or not 😅 

Me either, it's often a bit tricky to work out when they're useful. In principle if we're doing potentially expensive things like block parsing on a component that might get rendered multiple times, I think it makes sense to add a `useMemo`. If I add a wrapper `findPostContentWrapper` function that calls `findPostContent` and log out the number of times it's called, in my local environment it looks like we call it 9 times (not sure how valid that is, since there's only 5 log lines there, but it at least suggests we might benefit from the useMemo caching 🤔):

<img width="224" alt="image" src="https://user-images.githubusercontent.com/14988353/191167470-d40e2f72-037f-4e90-9e6c-4b473df1b955.png">

From memory we can't call a hook from within another hook, so the optimisation might only be to wrap the block containing the two lines `parse` and `findPostContent` in a useMemo, so that we only parse / find blocks when the template content changes?

But this can totally be something to look at in the final stages of the PR.

tellthemachines replied:
Ok I tried what you recommended and it seems to be working well! We can still see a jump in layout when the page first loads, I think because the initial data from `useSelect` takes a while to appear, but I don't think there's much to be done about that.

andrewserong replied:
> I think because the initial data from useSelect takes a while to appear, but I don't think there's much to be done about that.

Probably outside the scope of this PR, but I wonder if there's a way to preload the data for this selector? I see that the Navigation block has a preloading function [here](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/packages/edit-navigation/src/utils/index.js#L16) and sets apiFetch to use it [here](https://github.com/WordPress/gutenberg/blob/e102e3f853ea2fc15d976e5a633b1c0668beed5d/lib/experimental/navigation-page.php#L168). There's similar logic in gutenberg_initialise_editor [here](https://github.com/WordPress/gutenberg/blob/c1cf175c962ac2239020b80fb6a02341f7205f0b/lib/experimental/editor-settings.php#L47), so I wonder if there's a PHP filter of some kind to add in a desired API endpoint to pre-fetch or something? Might be a bit of a rabbithole 😅 

I don't at all understand the logic there, but just in case something along those lines sparks any ideas 🙂

andrewserong commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -82,18 +85,31 @@ function MaybeIframe( {
+			return blocks[ i ];
+		}
+		if ( blocks[ i ].innerBlocks.length ) {
+			return findPostContent( blocks[ i ].innerBlocks );
```
I _think_ we only want to `return` here if we've found a post content block? Otherwise it looks like `findPostContent` bails early without having found something if the first depth-first discovery does not match a post content block.

In my current template, it bails after examining the deepest separator block, rather than moving up a level to eventually find the post content block.

In the below screenshot, I logged out `looking at` for each instance of the for loop, and it appears to stop at the `core/separator`, where it should have moved up a level and found `core/spacer` and then moved down one to `core/post-content`.

<img width="1113" alt="image" src="https://user-images.githubusercontent.com/14988353/191151946-856009d5-c173-4d19-98c9-913cacf54555.png">

So, maybe something like the following might work?

```js
let foundPostContent;

...

foundPostContent = findPostContent( blocks[ i ].innerBlocks );
if ( foundPostContent ) {
    return foundPostContent;
}
```

tellthemachines replied:
Ahhhh well spotted! Yeah it'll have to be something along those lines.

andrewserong reviewed (commented):
I love the direction here @tellthemachines, this looks very promising!

When testing manually in my local environment, it looks like it wasn't able to find a post content block, even though `templateBlocks` contains a nested one:

<img width="991" alt="image" src="https://user-images.githubusercontent.com/14988353/191151349-0e15d4ca-9209-4c2e-a1b9-acd1ee39d86d.png">

I've left a comment on `findPostContent` — I'm pretty sure it might be a fairly simple change in that function to get it working.

Happy to do further testing!

tellthemachines committed:
Fix post content finding logic.

andrewserong commented:
Nice one, that's fixed up locating the `postContentBlock` 👍

I noticed that if I go to make a change in the post content block of the current template it crashes the editor. Not too sure why, though!

![2022-09-20 14 58 38](https://user-images.githubusercontent.com/14988353/191171534-72fe2053-4532-49c8-b1af-d1aae87a58fa.gif)

tellthemachines committed:
Remove extra styles when Post Content block exists.

tellthemachines committed:
templateContent should always be a string

tellthemachines committed:
Memoise post content block.

andrewserong commented:
I think I found another instance of `templateContent` oddly being not-a-string. If I go to update the Post Content block in the currently selected template in the template editor, and then close out of the template editor instead of saving, then the editor also crashes with the following error: (looks like it's returning a callback function instead of a string?)

| Error message | Clicking back from template editor |
| --- | --- |
| <img width="741" alt="image" src="https://user-images.githubusercontent.com/14988353/191408122-881fa0fd-7acd-4ccd-948c-8c08fc03c7d2.png"> | ![2022-09-21 13 25 25](https://user-images.githubusercontent.com/14988353/191408431-d57a31bd-72bd-4f32-a852-e17b79a226f4.gif) |

andrewserong commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -183,25 +205,50 @@ export default function VisualEditor( { styles } ) {
+	// if it doesn't have one. If no Post Content this must be a classic theme,
+	// in which case we use the fallback layout.
+	const postContentLayout = postContentBlock
+		? postContentBlock?.attributes?.layout || { type: 'default' }
```
For this one, do we need the fallback logic for legacy markup (e.g. `inherit`) here, too? I noticed in TwentyTwentyTwo, the `single` template appears to still be using `inherit` so the view in the post editor for me was full width because it was assuming `default` (flow) layout instead of `constrained`:

![image](https://user-images.githubusercontent.com/14988353/191408980-d841e9c4-d66d-4052-903e-d1ada78000f5.png)

tellthemachines replied:
Ooooh good point! On it.

andrewserong commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -286,9 +356,9 @@ export default function VisualEditor( { styles } ) {
 								className={
 									isTemplateMode
 										? 'wp-site-blocks'
-										: `${ blockListLayoutClass } wp-block-post-content` // Ensure root level blocks receive default/flow blockGap styling rules.
```
I was wondering if we might still need the `wp-block-post-content` classname to be added, so that we can hook into styles that are set at the `core/post-content` block in global styles? (It looks like the layout classnames don't output the block classname, so I think we might still need to add it in manually 🤔)

tellthemachines replied:
I'm a bit hazy on how the global styles stuff works in the post editor, so if you think it's needed I'll just add it back. Won't do any harm to have the class there 😄

andrewserong replied:
> I'm a bit hazy on how the global styles stuff works in the post editor

I was looking at this the other day while reviewing other PRs, so in case the breadcrumbs here are helpful: it's pretty naive when it comes to global styles. The server-rendered global styles are added to the block editor settings in PHP [here](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.1/block-editor-settings.php#L77), and then the post editor figures out whether or not to use them [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/editor.js#L167). The styles are then passed down to the post editor's `Layout` component [on this line](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/editor.js#L201). That component then passes the styles to `VisualEditor` [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/components/layout/index.js#L244). And in `VisualEditor`, the styles are passed to `MaybeIframe` which finally passes it along to `EditorStyles`, which ultimately outputs styles in a `style` tag on this line: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/editor-styles/index.js#L81

The TL;DR is that it's _nearly_ as if the global styles stylesheet had been enqueued, only those styles get transformed a bit by the `EditorStyles` component to play nicely in the editor.

I mostly just wanted to write that out somewhere because it finally clicked for me how it's hooked together 😀

tellthemachines replied:
Thanks, that's really helpful!

andrewserong reviewed (commented):
This is working pretty well so far, I'm really excited about the progress here!

I was wondering if it's also worth us hooking in the style engine to gather CSS for the other block supports that might be in use in the Post Content block instance in the current template. I'm mostly thinking of Typography supports like font-size, font family, and the like. It'd be cool if we could output those styles in addition to the layout styles (but I also totally understand if you think that's better for a separate PR). We might be able to get away with passing the block instance's style object to this function to get the CSS we need:

https://github.com/WordPress/gutenberg/blob/5fbd554da9ca8d23fca605d859a737cbb8c18513/packages/style-engine/src/index.ts#L27

Aside from the other couple of comments, I ran into another issue with TwentyTwentyTwo which is that it looks like its editor styles appear to override the layout content justification rules (it renders correctly on the site frontend). If I set the Post Content block in my current template to left content justification with custom content and wide sizes, here's how the Cover block looks in the editor:

![image](https://user-images.githubusercontent.com/14988353/191410455-1dee0fae-3e0c-4871-90ec-863d3bbd5c93.png)

tellthemachines committed:
Change function signature and mark experimental

tellthemachines committed:
_Really_ make sure templateContent is a string

tellthemachines committed:
Update layout type for legacy layouts.

tellthemachines commented:
Thanks for the continued feedback @andrewserong! I think I've addressed everything so far; the only remaining todoes are adding tests for the experimental functions we're exposing and working out why e2e are failing on this branch 🤔 

> I was wondering if it's also worth us hooking in the style engine to gather CSS for the other block supports that might be in use in the Post Content block instance in the current template. I'm mostly thinking of Typography supports like font-size, font family, and the like.

Definitely let's do that, but I'd rather it be a separate PR to keep things simple for this one. I'm also unsure of the urgency of that work 6.1-wise, as so far we've only had reports about the missing layout styles.

andrewserong commented:
Sounds good 👍

> I'm also unsure of the urgency of that work 6.1-wise, as so far we've only had reports about the missing layout styles.

I imagine it's most likely lower priority because themes can target `core/post-content` in global styles, so it's probably much rarer to need to set Typography rules at the individual block level. I think it'd only really make sense if your theme had one set of Typography styles for one template, and another for another template. E.g. if a site decided that their blog was in a mono-spaced typewriter-like font, but they wanted their About pages in a serif-font, that kinda thing.

andrewserong commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -183,25 +206,59 @@ export default function VisualEditor( { styles } ) {
+
+	const postContentLayoutStyles = useLayoutStyles(
+		postContentBlock,
+		'.wp-container-visual-editor'
```
I'm wondering if this selector needs to be a higher specificity selector such as `.block-editor-block-list__layout.is-root-container`, which I think was used prior to this PR? If I use that one on this line, the layout styles appear to win out over the TwentyTwentyTwo styles as expected. But I wasn't sure if you changed this selector for another reason?

| Current PR (TwentyTwentyTwo's margins override the left content justification) | When updating to the higher specificity selector |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/191433182-b2e48cd0-4f2d-4ba6-852c-fd170e53c54e.png) | ![image](https://user-images.githubusercontent.com/14988353/191432994-79fa6917-4d07-4f44-b26f-b2a374ab6f30.png) |

tellthemachines replied:
Not really! Must be my subconscious drive to lower specificity everywhere 😅 
It's a good idea to use the selector we had before though.

andrewserong commented on packages/block-editor/src/hooks/layout.js:
```diff
@@ -37,58 +37,105 @@ import { getLayoutType, getLayoutTypes } from '../layouts';
+	const { attributes = {}, name } = block;
+	const { layout = {}, style = {} } = attributes;
+	// Update type for blocks using legacy layouts.
+	const usedLayout =
+		layout &&
+		( layout?.type === 'constrained' ||
+			layout?.inherit ||
+			layout?.contentSize ||
+			layout?.wideSize )
+			? { ...layout, type: 'constrained' }
+			: { ...layout, type: 'default' };
```
Will we ever need to use this hook for flex layouts? With the current version, it looks like `layout.type` set to `flex` would result in output for the default layout.

tellthemachines replied:
Ooooooh right, that needs to change

tellthemachines commented:
Looking into adding tests for `useLayoutClasses` and `useLayoutStyles`, both of these use hooks internally that will have to be mocked. I wonder how useful it is to add these tests right now 🤔 as I suspect most of the bugs will come from wrong assumptions regarding what we pass into them, or what we retrieve with the hooks, rather than the actual logic in the functions which is pretty straightforward.

andrewserong reviewed (commented):
Thanks for all the updates again @tellthemachines, this feels much closer. I left a comment where I think we might be able to fix up the styling issue with the Cover block in TwentyTwentyTwo.

It looks like there's still an issue with going to the template editor and then exiting out after making an unsaved change. While the editor no longer crashes, it looks like only dealing with a string value results in the editor forgetting which layout type to render. In my local env, after making a change and exiting out, the layout defaults to `flow` again stretching everything full width:

![2022-09-21 16 49 26](https://user-images.githubusercontent.com/14988353/191434892-dfe9c0de-0101-45cd-83f0-198d0ab7c1e3.gif)

andrewserong commented:
> Looking into adding tests for useLayoutClasses and useLayoutStyles, both of these use hooks internally that will have to be mocked. I wonder how useful it is to add these tests right now

I agree, so long as we're manually testing each of the cases we're dealing with in these hooks, I think we can leave writing tests for them to a follow-up if need be 👍

tellthemachines committed:
Fix selector specificity and legacy layout logic.

tellthemachines committed:
Fix 404s on fetch and only parse content if needed

tellthemachines commented:
> It looks like there's still an issue with going to the template editor and then exiting out after making an unsaved change. While the editor no longer crashes, it looks like only dealing with a string value results in the editor forgetting which layout type to render. In my local env, after making a change and exiting out, the layout defaults to flow again stretching everything full width:

This is now fixed, and the post editor should show the most recent version of layout even if it wasn't saved in the template editor (I think this makes sense because it's still possible to save the template after exiting template editing mode).

I also fixed the failing e2e tests and addressed remaining feedback, so this is ready for another round of review/testing! For reference, I've been testing with:

* block themes
* classic themes with theme.json
* classic themes without theme.json

andrewserong reviewed (commented):
This is getting closer each time @tellthemachines!

I think I found another editor fatal. Using a template that has customizations made to it, if I go to open the template editor and then clear the customizations, this appears to throw a fatal, but doesn't on trunk:
 
![2022-09-26 12 25 16](https://user-images.githubusercontent.com/14988353/192182116-14605030-f48c-479f-8731-8f49b8cb3abb.gif)

I'm not too sure where this is happening yet, but just thought I'd share the screenshot as I found it. Happy to do a bit more investigation later on today. In case it helps, my test post content includes Buttons, Search block, and Cover blocks, though I doubt that's related.

ramonjd commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -82,20 +85,40 @@ function MaybeIframe( {
 	);
 }
 
+function findPostContent( blocks ) {
```
Could we help our future selves here and add a JS doc comment?

ramonjd commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -183,25 +210,69 @@ export default function VisualEditor( { styles } ) {
+	const postContentBlock = useMemo( () => {
+		// When in template editing mode, we can access the blocks directly.
+		if ( editedPostTemplate?.blocks ) {
+			return findPostContent( editedPostTemplate?.blocks );
```
I can reproduce error that @andrewserong  [noticed](https://github.com/WordPress/gutenberg/pull/44258#pullrequestreview-1119518999) but only sometimes. 

Most of the times it works for me. 

When it does fail _I suspect_ it's because postContentBlock is undefined (and therefore trying to access the attributes below fails).

Replacing with `return findPostContent( editedPostTemplate?.blocks ) || {};` seems to get around that, 

![2022-09-26 12 39 39](https://user-images.githubusercontent.com/6458278/192183452-1d765f0a-9f84-48f2-be7d-77e848d27384.gif)

But it exposes another error triggered by the <BlockPopoverInbetween> component 🤣 

`Uncaught TypeError: Cannot read properties of null (reading 'removeEventListener')`

https://github.com/WordPress/gutenberg/blob/ed8d25e651f9673583ae1c4800339eec45684830/packages/block-editor/src/components/block-popover/inbetween.js#L229

Not sure why yet. I can't reproduce either error in trunk.

tellthemachines replied:
Hmm, I've only managed to repro the error once on TT2, and not at all on Empty theme. But I think something's messed up because previous customisations are now appearing as the default template state 😅 
Will continue investigating.

tellthemachines replied:
Ok I fixed the failure if Post Content block is removed from the template by accessing layout from attributes via optional chaining instead of destructuring. The reason returning empty object from `findPostContent` won't work correctly is some logic further down is dependent on checking if postContentBlock exists, and empty object will always be `true` 😅

ramonjd reviewed (commented):
A bit of a fly-by review, sorry! I don't have a lot of context, but the code reads well and changes I make to the post content block in the singular template are reflect in the post editor. 👍 

Not sure what's going on with the inline template editing errors yet... 🤔

tellthemachines committed:
Fix acces to layout when no Post Content block

tellthemachines committed:
Add doc comment to findPostContent

tellthemachines commented:
> I think I found another editor fatal. Using a template that has customizations made to it, if I go to open the template editor and then clear the customizations, this appears to throw a fatal, but doesn't on trunk:

I managed to reproduce the issue on trunk with Empty theme. Steps to reproduce:

1. Create a new post and add some content but don't save or publish yet;
2. Go into post template editing mode and make a change to the template;
3. Save and publish both the template changes and the post;
4. Click "Clear customizations" in the top area.

I'm not yet sure what's behind it; the error is due to something null being passed to either `BlockParentSelector` or `BlockPopoverInbetween` (in my testing errors alternate between the two 🤔)
In any case it seems unlikely to be related to this PR.

andrewserong commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -183,25 +218,68 @@ export default function VisualEditor( { styles } ) {
-		'is-layout-flow': ! themeSupportsLayout,
-		'has-global-padding': useRootPaddingAwareAlignments,
-	} );
+	const postContentLayoutStyles = useLayoutStyles(
+		postContentBlock,
+		'.block-editor-block-list__layout.is-root-container'
+	);
```
I think there appears to be a difference in the rules that are being output here versus trunk 🤔 I noticed that there's an issue with the middle content justification with the Cover block set to wide alignment in TwentyTwentyTwo:

| Trunk | This PR |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/192202854-7f45060a-745c-48a8-9434-c3ad6dccd1d5.png) | ![image](https://user-images.githubusercontent.com/14988353/192203014-68708d70-0111-4572-a7b0-e0ce7219a989.png) |

This appears to be the rule that isn't present / winning out:

<img width="425" alt="image" src="https://user-images.githubusercontent.com/14988353/192203888-6d2f297d-a645-4b28-8581-d6565c6d317c.png">

andrewserong replied:
I wonder if the issue is that `useLayoutStyles` generates the layout styles for the individual post content block, but not the base layout styles that need to be attached to this more specific selector in order for it to win out? Or to put it differently, prior to this PR, it looks like we might have depended on the output of base layout rules with this more specific selector for this middle content justification. Apologies if I'm way off!

tellthemachines replied:
Those styles that are overriding the middle justification are the TT2 root padding logic that I thought were going to be removed after the Core root padding styles became available.

I'm reluctant to either increase specificity of the default layout styles or add `wp-container` rules for the default justification state, because any theme can conceivably still override them with higher specificity rules 😅 

This should probably be fixed in TT2 instead cc @carolinan @scruffian

andrewserong replied:
> I'm reluctant to either increase specificity of the default layout styles or add wp-container rules for the default justification state

Agreed, it'd be better not to have to add container rules, or explicit rules for the default justification state 🤔

> This should probably be fixed in TT2 instead

It's a tricky one... because TT2 has been used as a base for many other themes, it might not just TT2 out in the wild that has this issue. (E.g. the rule appears to be part of BlockBase https://github.com/Automattic/themes/blob/23b9090090279c95788357bb634fd2e57c3973db/blockbase/sass/base/_alignment.scss#L58 which many themes have been built on top of, too).

Just trying to think of other options... would it work to update the change to this line in this PR https://github.com/WordPress/gutenberg/blob/67c829dc9d4a265c22961623c02c42185321c2b2/packages/edit-post/src/components/visual-editor/index.js#L340 to include `.block-editor-block-list__layout.is-root-container` as well (as on trunk)? Or did we we need to remove it for the `postContentLayoutStyles` to work? Since they already have the selector attached, I was wondering if they can coexist. From quickly adding it back in in my local environment, it appears to get all three content justifications working in TT2 for me, but it's highly likely I'm missing some context 😅

tellthemachines replied:
> Just trying to think of other options... would it work to update the change to this line in this PR to include `.block-editor-block-list__layout.is-root-container` as well (as on trunk)? Or did we we need to remove it for the `postContentLayoutStyles` to work?

I did remove it initially so that custom content widths would work, but because we since increased the specificity of the default rules it seems to be working ok. I just tried pushing that change now, hopefully it doesn't break anything else 😅  

> It's a tricky one... because TT2 has been used as a base for many other themes, it might not just TT2 out in the wild that has this issue.

I'm not sure how much of a concern that should be to us, because the temporary status of these rules is explicitly called out in the TT2 [stylesheet](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-content/themes/twentytwentytwo/style.css#L87).

andrewserong reviewed (commented):
> I'm not yet sure what's behind it; the error is due to something null being passed to either BlockParentSelector or BlockPopoverInbetween (in my testing errors alternate between the two 🤔)
> In any case it seems unlikely to be related to this PR.

Thanks for looking into that! One of the usages of `BlockPopoverInbetween` is in the [`ZoomOutModeInserters`](https://github.com/WordPress/gutenberg/blob/c1cf175c962ac2239020b80fb6a02341f7205f0b/packages/block-editor/src/components/block-tools/zoom-out-mode-inserters.js#L39) — I noticed that if I switch off that experimental feature, I can't seem to replicate the issue immediately, so perhaps it's related to that implementation? In any case, I agree, it sounds like it's unrelated to this change, maybe we just noticed it here.

Separately, I think I found another styling rule issue with the middle content justification, so I've added another comment. Not too sure if it's the right place?

tellthemachines committed:
Increase specificity

andrewserong commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -257,13 +335,36 @@ export default function VisualEditor( { styles } ) {
+											}
+										/>
+									) }
+									{
+										// For classic themes using theme.json we want a default content width in the editor.
+										! postContentBlock && (
+											<LayoutStyle
+												selector=".block-editor-block-list__layout.is-root-container"
+												layout={ fallbackLayout }
+												layoutDefinitions={
+													globalLayoutSettings?.definitions
+												}
+											/>
+										)
 									}
```
Now that the above `<LayoutStyle>` also includes `.block-editor-block-list__layout.is-root-container`, it looks like we no longer need this conditional for classic themes, because it'll already be covered by the one above?

tellthemachines replied:
Oh yes, well spotted!

andrewserong reviewed (approved):
Nice one @tellthemachines, I think adding that rule back in did the trick. This is all testing nicely for me now. Also, adding back that rule seems to have fixed a flash of full width alignment on blocks when using the centre content justification in the post editor, so the UI feels a little more stable on a full page reload to me, now 👍 

Tested with the following:

✅ TwentyTwentyTwo with left, center, right alignments, with a variety of content and wide size widths
✅ Tove blocks-based theme
✅ Emptytheme with the above, and `useRootPaddingAwareAlignments` to check that the new global padding settings still work correctly
✅ Classic themes TT1, TT, TwentyNineteen

I just left a comment where I think we can now remove the `<LayoutStyle>` component for classic themes since it'll already be covered in the earlier instance. But otherwise, this LGTM! ✨

tellthemachines committed:
Remove unneeded LayoutStyle

tyxla commented:
Just closing the loop that in #44506 we addressed a performance regression introduced by this PR.

michalczaplinski commented:
Cherry-picked for the 6.1 release in https://github.com/WordPress/gutenberg/pull/44656

youknowriad commented on packages/edit-post/src/components/visual-editor/index.js:
```diff
@@ -183,25 +218,68 @@ export default function VisualEditor( { styles } ) {
+	// if not, this must be a classic theme, in which case we use the fallback layout.
+	const blockListLayout = postContentBlock
+		? postContentLayout
+		: fallbackLayout;
```
Hi there! I stumbled on this code while doing some performance debugging. While right now, it's not clear to me that this code (I mean all the figuring out of the layout) is introducing performance regression (specially loading) but I noticed that the value of the layout changes three times for a regular post loaded in the post editor.

<img width="593" alt="Screen Shot 2022-10-24 at 11 08 48 PM" src="https://user-images.githubusercontent.com/272444/197640055-6f3f69a1-454a-46fc-a646-03350a80309c.png">

- A few things stand out to me here. Can we just compute the right layout only once to avoid unnecessary re-renders?
- Why does the layout object contains this "definitions" key? Seems weird? (probably unrelated to this PR but maybe you have an idea)
- Also the logic here is over-complexifying this component. Is there a possibility of some kind of cleanup like a dedicated hook `useBlockEditorLayout` or something (I have no idea, just wild suggesting)?

Thanks

tellthemachines replied:
We don't have access to the post template straightaway on load, so we need to do a little checking to see when it becomes available. But we should probably be checking whether postContentBlock is an empty object rather than just checking whether it exists, as then `blockListLayout` would only change once.

The definitions object appeared in #40875; it's where we store base styles for each layout type. When we get the layout settings with `useSetting` the definitions will always be included.

>  Is there a possibility of some kind of cleanup like a dedicated hook useBlockEditorLayout or something

That's a good point, I can't say offhand but it's worth looking into!

youknowriad replied:
> The definitions object appeared in https://github.com/WordPress/gutenberg/pull/40875; it's where we store base styles for each layout type. When we get the layout settings with useSetting the definitions will always be included.

What's seems not correct here is that the "definition" are included in the "default layout" object. I'm fine having "definitions" but I don't they should be part of the default layout object.

youknowriad replied:
>  we should probably be checking whether postContentBlock is an empty object rather than just checking whether it exists, as then blockListLayout would only change once.

Yeah, is there a way to have no changes at all, this is to avoid rendering anything if the correct layout object is not available yet. As I guess otherwise, we'd just be doing unnecessary computation and potentially causing layout shifts in the UI.

tellthemachines replied:
> I'm fine having "definitions" but I don't they should be part of the default layout object.

I'm not sure I understand. The default layout type has default styles too, so we need to access those definitions sometimes. Do you mean we shouldn't be returning the definitions from `useSetting( 'layout' )` or am I missing your point altogether? 😅 

> Yeah, is there a way to have no changes at all, this is to avoid rendering anything if the correct layout object is not available yet. As I guess otherwise, we'd just be doing unnecessary computation and potentially causing layout shifts in the UI.

Yeah, there will be a layout shift if the layout is anything other than center-aligned and constrained to default theme width. The problem is, if we wait for the post template data to be available before rendering the post content, users will be looking at a white screen for a little bit, which seems to me a worse experience than the layout shift. Unless we can find a faster way to access the Post Content block layout? The thing is we need the Post Content block from the specific template being used with the post, because it might have custom settings. It's not enough to get the data from global styles.

tellthemachines replied:
Thinking through this a bit, also in light of #45262, would it be terrible to get the Post Content layout and add it to the block editor settings[ here](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.1/block-editor-settings.php#L15)? I haven't tried it out yet but happy to do so as I think we're going to need a solution that renders the correct layout for all types of users, not just the ones with the right permissions 😬

youknowriad replied:
I think in general we're trying to move away from server-side provided settings in order to make the editor more portable (outside wordpress, other contexts...) but in this case, I wouldn't mind it. The original idea of the "layout" properly in theme.json was to avoid this problem exactly: the layout in theme.json would always correspond (as a convention) to the post content block one and we'd avoid any complex computations. (statically defined)

tellthemachines replied:
Thanks for the context! Yeah, the issue is precisely that folks wanted to customize the Post Content layout, and with features such as #44065 it makes sense for the post editor to reflect whatever layout is set.
